### PR TITLE
[pipelines] skipping underwater accounts [GEN-8490]

### PR DESCRIPTION
**Pull request by ochaloup**
There are failures when funding settlements because the funding program cannot handle partial funding of undelegated stake accounts. Such accounts have to be skipped in the pipeline.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved settlement funding reliability by excluding stake accounts that could trigger underflow or splitting issues during funding selection.
  * Added clearer warnings to flag delegation amounts associated with accounts that may be skipped.
  * Ensures eligible stake accounts remain available for other settlement funding plans.
* **Tests**
  * Added unit test coverage for underflow/splitting edge cases, including scenarios where an “underwater” account is the only candidate.

### CODE REVIEW

**Review by martinkrecek — DISMISSED** [review]
The current open comment is worth a double-check, I'm not sure if some long-term issues can appear or if it will be funded at some next attempt.

File: settlement-pipelines/src/bin/fund_settlement.rs [hunk 1]
```diff
@@ -345,6 +346,31 @@ async fn prepare_funding(
...
+                                reason_display(&settlement_record.reason),
+                                account.stake_account,
+                                build_balance_message(account.lamports, false, false),
+                                build_balance_message(delegation_stake, false, false),
+                            )).with_vote(settlement_record.vote_account_address).add();
+                            false
+                        }
+                        None => true,
+                    }
+                });
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply settlement-pipelines/src/bin/fund_settlement.rs]: 🤖 claude: Fixed locally (not pushed yet).

Confirmed, and the impact was slightly worse than a deferral: `funding_stake_accounts` is a `&mut Vec` borrowed straight out of the shared `fund_bond_stake_accounts` map, and the settlement loop iterates a `HashMap<epoch, Vec<SettlementRecord>>`, so which settlement got to delete the account first varied between runs. Since `splits` goes *false* as `amount_needed` grows, the account dropped while serving a small settlement is exactly the one a larger settlement could have consumed safely on the whole-account path.

What changed in `prepare_funding`:

- `underwater_split_stake_accounts()` only *identifies* the affected accounts (and feeds the per-account warnings) — it no longer mutates the pool.
- `take_stake_accounts_to_fund()` keeps them in the pool and simply does not select them for the current settlement; only accounts actually taken for funding are removed.
- the `Available N stake accounts (...)` log line now excludes the skipped ones, so it reports what this settlement can really draw on.

Regression test added as requested — `underwater_account_skipped_for_splitting_settlement_stays_available_for_a_bigger_one`: one pool, two settlements. A 10 SOL account with a stale 20 SOL delegation is skipped for `amount_needed = 3 SOL` (split path) while its healthy sibling funds that settlement, stays in the pool, and is then selected for `amount_needed = 10 SOL`, where the program takes the whole-account branch and never touches `delegation.stake`. A second test pins the safety side: the account is never selected for the splitting settlement even when it is the only candidate. Both tests fail against the previous `retain` version — verified by temporarily reintroducing the removal.

**Comment by ochaloup** [comment]
🤖 claude: Review triage for this PR. Code is fixed.

| Finding | Disposition |
| --- | --- |
| CodeRabbit inline: underwater accounts permanently removed from the shared pool | FIXED — details in the [thread reply](https://github.com/marinade-finance/validator-bonds/pull/479#discussion_r3656191749) |
| Two-settlement regression test missing (also raised in my own round-1 branch review: filter behavior untested, only the pure helper covered) | FIXED — 2 tests added in `fund_settlement.rs`, both fail against the previous code |
| @martinkrecek: "worth a double-check, I'm not sure if some long-term issues can appear or if it will be funded at some next attempt" | ANSWERED — below |
| Pre-merge check "Ticket Reference In Pr Title" | DECLINED — see below |

**Review by martinkrecek — APPROVED** [review]
